compiler.go

```diff
@@ -890,6 +922,187 @@ found:
+	numVars := len(scope.bindings) - numFuncs
+	vars := make([]unistring.String, len(scope.bindings))
+	for i, b := range scope.bindings {
+		vars[i] = b.name
+	}
+	if len(vars) > 0 && !ownVarScope && ownLexScope {
+		if inGlobal {
+			c.emit(&bindGlobal{
+				vars:      vars[numFuncs:],
+				funcs:     vars[:numFuncs],
+				deletable: eval,
+			})
+		} else {
+			c.emit(&bindVars{names: vars, deletable: eval})
+		}
+	}
+	var enter *enterBlock
+	if c.compileLexicalDeclarations(in.Body, ownVarScope || !ownLexScope) {
+		if ownLexScope {
+			c.block = &block{
+				outer:      c.block,
+				typ:        blockScope,
+				needResult: true,
+			}
+			enter = &enterBlock{}
+			c.emit(enter)
+		}
+	}
+
+	for _, exp := range in.Body {
+		if imp, ok := exp.(*ast.ImportDeclaration); ok {
+			c.compileImportDeclaration(imp)
+		}
+	}
+	for _, entry := range module.localExportEntries {
+		name := unistring.NewFromString(entry.localName)
+		b, ok := scope.boundNames[name]
+		if !ok {
+			if entry.localName != "default" {
+				// TODO fix index
+				c.throwSyntaxError(0, "exporting unknown binding: %q", name)
+			}
+			b, _ = scope.bindName(name)
+		}
+
+		b.inStash = true
+		b.markAccessPoint()
+
+		exportName := unistring.NewFromString(entry.localName)
+		callback := func(vm *vm, getter func() Value) {
+			vm.r.modules[module].(*SourceTextModuleInstance).exportGetters[exportName.String()] = getter
+		}
+
+		if entry.lex || !scope.boundNames[exportName].isVar {
+			c.emit(exportLex{callback: callback})
+		} else {
+			c.emit(export{callback: callback})
+		}
+	}
+	for _, entry := range module.indirectExportEntries {
+		otherModule, err := c.hostResolveImportedModule(c.module, entry.moduleRequest)
+		if err != nil {
+			panic(fmt.Errorf("previously resolved module returned error %w", err))
+		}
+		if entry.importName == "*" {
+			continue
+		}
+		b, ambiguous := otherModule.ResolveExport(entry.importName)
+		if ambiguous || b == nil {
+			c.compileAmbiguousImport(unistring.NewFromString(entry.importName))
+			continue
+		}
+
+		exportName := unistring.NewFromString(entry.exportName).String()
+		importName := unistring.NewFromString(b.BindingName).String()
+		c.emit(exportIndirect{callback: func(vm *vm) {
+			m := vm.r.modules[module]
+			m2 := vm.r.modules[b.Module]
+			m.(*SourceTextModuleInstance).exportGetters[exportName] = func() Value {
+				return m2.GetBindingValue(importName)
+			}
+		}})
+	}
+	if len(scope.bindings) > 0 && !ownLexScope {
+		var lets, consts []unistring.String
+		for _, b := range c.scope.bindings[numFuncs+numVars:] {
+			if b.isConst {
+				consts = append(consts, b.name)
+			} else {
+				lets = append(lets, b.name)
+			}
+		}
+		c.emit(&bindGlobal{
+			vars:   vars[numFuncs:],
+			funcs:  vars[:numFuncs],
+			lets:   lets,
+			consts: consts,
+		})
+	}
+	if !inGlobal || ownVarScope {
+		c.compileFunctions(funcs)
+	}
+	// TODO figure something better 😬
+	c.emit(notReallyYield) // this to stop us execute once after we initialize globals
```

Comment:
This likely is the most hacky part of the whole PR. Yes A lot of the others might not be the greatest code I have ever written, or might need some refactoring, but at least I am pretty certain it does what it's supposed to. [NotReallyYield](https://github.com/mstoykov/goja/blob/5f81441173723ebb181873312ffaa98604cdd2b0/vm.go#L1437-L1455) though is ... different. This exists as by the specification all the definitions need to have happened in a module in the [InitializeEnvironment](https://262.ecma-international.org/13.0/#sec-source-text-module-record-initialize-environment) which happens in the [innerModuleLinking](https://262.ecma-international.org/13.0/#sec-InnerModuleLinking) which is in practice needs to be part of the "compilation".

But those defintions are in practice the first things that are executed (as you well know) so I in practice needed a way to start executing the module code ... get through the definition and then *stop*, and go actually execute the code of a dependancy module. And then later continue from that [place](https://github.com/mstoykov/goja/blob/5f81441173723ebb181873312ffaa98604cdd2b0/compiler.go#L1093-L1094) - preferably with the same stack as otherwise I will need to change all of the code about where does goja put variables.

This sounds awful lot as yield, so this is what I "implemented". I did think I will need more stuff but ... not really it's literally one instruction and (because I didn't want to go down the rabbit hole of implementing the whole generator stuff ...) a way to [restart execution](https://github.com/mstoykov/goja/blob/5f81441173723ebb181873312ffaa98604cdd2b0/runtime.go#L1436).

I did want to try to implement generators at somepoint, but priorities change, so I am pretty sure it will take me some time (at least 1-2 months) to get around to it. And we do definitely want to have `async`/`await` support and IIRC that was the primary blocker, so I will likely get on with it at some point in the future. And from my quick try - there are bunch of other stuff I will need to do to make real generators work, most of which hopefully will be just reusing code you already wrote - but some won't be ;(.

But back to this notReallyYield - I did test it a bunch and found some problems with the original way I wrote it, but this version seems to work really well. I still expect it's likely buggy just not in a way I can figure out, so definitely look at it closely and maybe a different approach will be nice 🤷‍♂